File: src/solutions/scala/N5LongestPalindromicSubstring.scala
package solutions.scala

import scala.annotation.tailrec

object N5LongestPalindromicSubstring {

  def longestPalindrome(s: String): String = {
    @tailrec
    def expandAroundCenter(left: Int, right: Int): Int = {
      if (left < 0 || right >= s.length || s.charAt(left) != s.charAt(right)) right - left - 1
      else expandAroundCenter(left - 1, right + 1)
    }

    @tailrec
    def longestPalindrome(i: Int, start: Int, end: Int): String =
      if (i >= s.length) s.substring(start, end + 1)
      else {
        val len = Math.max(expandAroundCenter(i, i), expandAroundCenter(i, i + 1))
        if (len > end - start) longestPalindrome(i + 1, i - (len - 1) / 2, i + len / 2)
        else longestPalindrome(i + 1, start, end)

      }

    Option(s)
      .filter(!_.isEmpty)
      .map(_ => longestPalindrome(0, 0, 0))
      .getOrElse("")
  }

  def main(args: Array[String]): Unit = {
    assert(Iterator("bab", "aba").contains(longestPalindrome("babad")))
    assert(longestPalindrome("cbbd") == "bb")
    assert(longestPalindrome("") == "")
  }
}

File: src/solutions/java/N5LongestPalindromicSubstring.java
package solutions.java;

public class N5LongestPalindromicSubstring {

    public String longestPalindrome(String s) {
        if (s == null || s.length() < 1) {
            return "";
        }
        int start = 0;
        int end = 0;
        for (int i = 0; i < s.length(); i++) {
            int len1 = expandAroundCenter(s, i, i);
            int len2 = expandAroundCenter(s, i, i + 1);
            int len = Math.max(len1, len2);
            if (len > end - start) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }
        return s.substring(start, end + 1);
    }

    private int expandAroundCenter(String s, int left, int right) {
        int l = left;
        int r = right;
        while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
            l--;
            r++;
        }
        return r - l - 1;
    }

    public static void main(String[] args) {
        N5LongestPalindromicSubstring app = new N5LongestPalindromicSubstring();
        System.out.println(app.longestPalindrome("babad"));
        System.out.println(app.longestPalindrome("cbbd"));
        assert app.longestPalindrome("babad").equals("bab") || app.longestPalindrome("babad").equals("aba");
        assert app.longestPalindrome("cbbd").equals("bb");
    }
}
